lib/Core/Memory.h

```diff
@@ -317,48 +321,159 @@ class PersistentState : public ObjectState {
     UpdateList cacheLineUpdates;
     UpdateList pendingCacheLineUpdates;
 
+    // (iangneal): We need a small symbolic array to construct an index variable which
+    // can represent a range.
+    UpdateList idxEmptyUpdates;
+    ref<Expr> idxUnbounded;
+    // We need to solve in slices for efficiency purposes.
+    const unsigned nrLinesPerSlice = 1; // pageszie
+    std::vector<ref<Expr>> idxConstraints;
+
+    /**
+     * (iangneal): We want symbolic root-cause detection. 
+     * 
+     * This is why:
+     * Each symbolic write can dirty a potentially large range of values if a
+     * symbolic offset is used as part of the write. At the extreme, the program
+     * could write at an unconstrained offset, potentially dirtying any 
+     * cacheline!
+     * 
+     * We already have facilities for solving whether or not a cacheline is 
+     * dirty or clean. However, for each cacheline we determine to not be clean,
+     * we would also like to know where it was most recently modified. This way,
+     * we can let the developer know where they might need to add some persists,
+     * or at least begin their search for finding the bug. Without this 
+     * information, it is nearly impossible to figure out where the bug is for
+     * programs of non-trivial size.
+     * 
+     * To track these root causes, we use a similar symbolic cacheline
+     * construction to how we track clean/dirty bits. Every time we dirty a 
+     * cache line, we update the state of the cacheline, and also write a 
+     * unique ID value into the root cause update list at the same symbolic 
+     * offset. When we perform a flush, we write a 0, equivalent to "no root
+     * cause." As a performance optimization, we only solve this array when we
+     * determine that the entire system is not guaranteed to be persistent.
```

Comment:
1. I'm actually thinking about moving the cacheLines array to Expr::Bool (one bit) as it is (It didn't work for me first try either, but there's some special handling code for booleans in ObjectState that I think we could use).
2. I only want to resolve this list lazily, so I don't want to assume the number of outstanding write id's is less than 256.
3. I'm concerned about the performance implications of having the array's have a larger range than 8 bits. 

What I will do is:
1. Measure the performance impact of having the cacheline array be larger. If it is negligible, I will merge these arrays.

Reply:
Okay, I'm on the same page. If the range of the array makes a difference, then it makes sense to have them separate, because we'll much more commonly query on the cacheLines array than the rootCause array.